File: typescript/src/Wall/WallService.ts
import {Brick} from './Brick';
import {WallDAO} from './WallDAO';
import {UsersAreNotFriendsError} from '../Error/UsersAreNotFriendsError';
import {UserNotLoggedInError} from '../Error/UserNotLoggedInError';
import {User} from '../User/User';
import {UserSession} from '../User/UserSession';

export class WallService {
    anotherBrickInTheWall(user: User, message: string): Brick[] {
        let wall: Brick[] = [];
        const loggedUser = this.getLoggedUser();
        let isFriend: boolean = false;

        if (loggedUser !== undefined) {
            for (const friend of user.getFriends()) {
                if (friend === loggedUser) {
                    isFriend = true;
                    break;
                }
            }

            if (isFriend) {
                wall = WallDAO.findBricksByUser(user);
                const brick: Brick = new Brick(message, new Date());
                WallDAO.addBrickToUser(user, brick);

                wall.push(brick);
                return wall;
            }

            throw new UsersAreNotFriendsError();
        } else {
            throw new UserNotLoggedInError();
        }
    }

    protected getLoggedUser(): User | undefined {
        return UserSession.getInstance().getLoggedUser();
    }
}

File: dotNet/src/Wall/WallService.cs
using src.Exception;
using src.Wrapper;

namespace src.Wall;

public class WallService
{
    private readonly IWallDAO _wallDao;
    private readonly IClock _clock;

    public WallService(IWallDAO wallDao, IClock clock)
    {
        _wallDao = wallDao;
        _clock = clock;
    }

    public List<Brick> AnotherBrickInTheWall(User.User user, string message, User.User? loggedInUser)
    {
        if (loggedInUser == null)
        {
            throw new UserNotLoggedInException();
        }

        if (!user.IsFriendWith(loggedInUser))
        {
            throw new UsersAreNotFriendsException();
        }

        List<Brick> wall = _wallDao.GetBricks(user);
        Brick brick = new Brick(message, _clock.Now());
        _wallDao.AddBrick(user, brick);

        wall.Add(brick);
        return wall;
    }
}
